modules/programs/fish.nix

```diff
@@ -74,7 +83,64 @@ in
   };
 
   config = mkIf cfg.enable {
-    home.packages = [ pkgs.fish ];
+    home.packages = [ cfg.package ];
+
+    xdg.dataFile."fish/home-manager_generated_completions".source =
+      let
+        # paths later in the list will overwrite those already linked
+        destructiveSymlinkJoin =
+          args_@{ name
+              , paths
+              , preferLocalBuild ? true
+              , allowSubstitutes ? false
+              , postBuild ? ""
+              , ...
+              }:
+          let
+            args = removeAttrs args_ [ "name" "postBuild" ]
+              // { inherit preferLocalBuild allowSubstitutes; }; # pass the defaults
+          in pkgs.runCommand name args
+            ''
+              mkdir -p $out
+              for i in $paths; do
+                if [ -z "$(find $i -prune -empty)" ]; then
+                  cp -srf $i/* $out
+                fi
+              done
+              ${postBuild}
+            '';
+        generateCompletions = package: pkgs.runCommand
+          "${package.name}-fish-completions"
+          {
+            src = package;
+            nativeBuildInputs = [ pkgs.python2 ];
+            buildInputs = [ cfg.package ];
+            preferLocalBuild = true;
+            allowSubstitutes = false;
+          }
+          ''
+            mkdir -p $out
+            if [ -d $src/share/man ]; then
+              find $src/share/man -type f | xargs python ${cfg.package}/share/fish/tools/create_manpage_completions.py --directory $out >/dev/null
+            fi
+          '';
+      in
+        destructiveSymlinkJoin {
```

Comment:
I merged as-is but you might want to check whether it would not be simpler/faster to just produce a single derivation having all the generated completions rather than combining the output of one derivation per package. That is, to have something like
```
xdg.dataFile."fish/home-manager_generated_completions".source =
  let
    cmd = ''
      mkdir $out
      ${concatMapStringsSep "\n" completionCmd config.home.packages}
    ''

    completionCmd = ''
      if [ -d $src/share/man ]; then
        find $src/share/man -type f -print0 \
          | xargs -0 python ${cfg.package}/share/fish/tools/create_manpage_completions.py --directory $out \
          > /dev/null
      fi
    '';
  in
    pkgs.runCommand
      "blahblah"
      { … }
      cmd;
```

Reply:
The problem with that would be that if any installed package changes all completions have to be regenerated, which can take a considerable amount of time if a lot of packages are installed.(exactly what this PR is supposed to avoid compared to just running `fish_update_completions` in a activation script)